test/refactoring/ConvertAsync/convert_bool.swift

```diff
@@ -0,0 +1,271 @@
+// BOOL-DONT-HANDLE-NEXT:   if <#err#> != nil {
+// BOOL-DONT-HANDLE-NEXT:     fatalError("oh no \(<#err#>!)")
+// BOOL-DONT-HANDLE-NEXT:   }
+// BOOL-DONT-HANDLE-NEXT: }
+// BOOL-DONT-HANDLE-NEXT: if !success {
+// BOOL-DONT-HANDLE-NEXT:   _ = <#err#> != nil ? fatalError("oh no \(<#err#>!)") : fatalError("some worries")
+// BOOL-DONT-HANDLE-NEXT: }
+// BOOL-DONT-HANDLE-NEXT: print("not err")
+
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path  | %FileCheck -check-prefix=BOOL-DONT-HANDLE2 %s
+boolWithErr { success, err in
+  if !success {
+    func doThings() {
+      fatalError("oh no \(err!)")
+    }
+    doThings()
+  }
+  if !success {
+    let doThings = {
+      fatalError("oh no \(err!)")
+    }
+    doThings()
+  }
+  if !success {
+    while err != nil {
+      fatalError("oh no \(err!)")
+    }
+  }
+  if !success {
+    for x: Int in [] {
+      fatalError("oh no \(err!)")
+    }
+  }
+  print("not err")
+}
+
+// BOOL-DONT-HANDLE2:      let success = try await boolWithErr()
+// BOOL-DONT-HANDLE2-NEXT: if !success {
+// BOOL-DONT-HANDLE2-NEXT:   func doThings() {
+// BOOL-DONT-HANDLE2-NEXT:     fatalError("oh no \(err!)")
+// BOOL-DONT-HANDLE2-NEXT:   }
+// BOOL-DONT-HANDLE2-NEXT:   doThings()
+// BOOL-DONT-HANDLE2-NEXT: }
+// BOOL-DONT-HANDLE2-NEXT: if !success {
+// BOOL-DONT-HANDLE2-NEXT:   let doThings = {
+// BOOL-DONT-HANDLE2-NEXT:     fatalError("oh no \(<#err#>!)")
+// BOOL-DONT-HANDLE2-NEXT:   }
+// BOOL-DONT-HANDLE2-NEXT:   doThings()
+// BOOL-DONT-HANDLE2-NEXT: }
+// BOOL-DONT-HANDLE2-NEXT: if !success {
+// BOOL-DONT-HANDLE2-NEXT:   while <#err#> != nil {
+// BOOL-DONT-HANDLE2-NEXT:     fatalError("oh no \(<#err#>!)")
+// BOOL-DONT-HANDLE2-NEXT:   }
+// BOOL-DONT-HANDLE2-NEXT: }
+// BOOL-DONT-HANDLE2-NEXT: if !success {
+// BOOL-DONT-HANDLE2-NEXT:   for x: Int in [] {
+// BOOL-DONT-HANDLE2-NEXT:     fatalError("oh no \(<#err#>!)")
+// BOOL-DONT-HANDLE2-NEXT:   }
+// BOOL-DONT-HANDLE2-NEXT: }
+// BOOL-DONT-HANDLE2-NEXT: print("not err")
+
+// err is not force unwrapped, so don't handle.
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path  | %FileCheck -check-prefix=BOOL-DONT-HANDLE3 %s
+boolWithErr { success, err in
+  if !success {
+    fatalError("oh no maybe \(err)")
+  }
+  print("not err")
+}
+
+// BOOL-DONT-HANDLE3:      let success = try await boolWithErr()
+// BOOL-DONT-HANDLE3-NEXT: if !success {
+// BOOL-DONT-HANDLE3-NEXT:   fatalError("oh no maybe \(<#err#>)")
+// BOOL-DONT-HANDLE3-NEXT: }
+// BOOL-DONT-HANDLE3-NEXT: print("not err")
+
+// RUN: %refactor -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):1 -I %S/Inputs -sdk %clang-importer-sdk-path  | %FileCheck -check-prefix=BOOL-WITH-ERR-SILLY %s
+boolWithErr { success, err in
+  if success == false && err == nil {
```

Comment:
I agree kind of silly, but this sort of implies that `success` is actually `error`. We *could* handle it as such... Thoughts?

Reply:
Yeah that's a fair point! It seems we should really make the determination of whether the flag is a success check or failure check from the branch in which the force unwrap occurs rather than the other way around. I think it would then also make sense to get rid of the bit of logic that discounts a flag condition if its path disagrees with the rest of the condition paths – I'll go ahead and make both those changes.

Reply:
Hmm, yeah we could handle that – it seems we'd want to change the condition path of the flag check if it disagrees with the rest of the conditions rather than dropping it entirely. I'll make that change.

On a related note, it also seems we should initially make the determination of whether the flag is a success check or failure check based on the branch in which the force unwrap occurs rather than the other way around, which would let us classify a flag as an `error` check if we then force unwrap the error.

Reply:
Hmm, the problem is that we're only classifying it as a success flag to begin with because of the force unwrap of the error in the body, so even if we classify both checks as success paths, the code still doesn't make much sense. We could apply a more general heuristic that allows for bool condition checks if they're also with other conditions that are able to classify as success or failure, but I don't really have a good sense of how often that comes up.

On a related note, I've just noticed that if the force unwrap of the error in this example was in an `else` block here, we would still classify the condition as the same, which doesn't seem right. It seems we should adjust the classification logic so we make the determination of whether the flag is a success check or failure check based on the branch in which the force unwrap of the error occurs rather than the other way around, which would let us classify a flag as an `error` check if we then force unwrap the error.

Reply:
Probably a couple of problems here really. Yes, ideally if the force unwrap was in the else we would then consider `success` to be a flag. Could get a bit tricky when the `else` can itself be another `if` though.

Another problem is that if we don't classify the `success` as being a flag then there are unhandled conditions and so we don't remove the `if` at all. Ideally we'd classify the whole if as being in the success block and then remove the `&& err == nil`. But that's an existing issue so... 🤷